### PR TITLE
ipn{,/ipnlocal}, client/tailscale: move Taildrop recv notifications to LocalAPI HTTP method

### PR DESCRIPTION
**Pull request by bradfitz**
Updates #6417

### CODE REVIEW

File: ipn/ipnlocal/local.go [hunk 1]
```diff
@@ -3558,6 +3562,47 @@ func (b *LocalBackend) WaitingFiles() ([]apitype.WaitingFile, error) {
 	return apiSrv.WaitingFiles()
 }
 
+// mapSetHandle is a minimal (but non-zero) value whose addresse serves as a map
```
Comment (andrew-d) [inline ipn/ipnlocal/local.go:3565]:
```suggestion
// mapSetHandle is a minimal (but non-zero) value whose address serves as a map
```

File: ipn/ipnlocal/local.go [hunk 2]
```diff
@@ -3558,6 +3562,47 @@ func (b *LocalBackend) WaitingFiles() ([]apitype.WaitingFile, error) {
...
+	b.setFileWaiter(handle, gotFileCancel)
+	defer b.setFileWaiter(handle, nil)
+
+	// After we've registered our waiter, check if anything's available. This
+	// guarantees there's no window for a file to arrive between checking and
+	// registering our waiter and us blocking forever.
+	go func() {
+		if ff, _ := b.WaitingFiles(); len(ff) > 0 {
+			gotFileCancel()
+		}
```
Comment (andrew-d) [inline ipn/ipnlocal/local.go:3595]: You could return the `ff` from `AwaitFile` here and simplify the `for` loop in `serveFiles`, I think?

Comment (bradfitz) [reply ipn/ipnlocal/local.go:3595]: I didn't really want to reason about the races. Getting the list is cheap. I just want some cheap sleep+wake-up call.

Would you prefer it returned nothing instead? Then caller can check its own ctx.Err() that it passed in.

Comment (bradfitz) [reply ipn/ipnlocal/local.go:3595]: Oh I think I know what you mean now. Let me try.

Comment (bradfitz) [reply ipn/ipnlocal/local.go]: PTAL

File: ipn/ipnlocal/local.go [hunk 3]
```diff
@@ -3551,13 +3555,63 @@ func (b *LocalBackend) TestOnlyPublicKeys() (machineKey key.MachinePublic, nodeK
...
+		return ff, err
+	}
+
+	for {
+		gotFile, gotFileCancel := context.WithCancel(context.Background())
+		defer gotFileCancel()
+
+		handle := new(mapSetHandle)
+		b.setFileWaiter(handle, gotFileCancel)
+		defer b.setFileWaiter(handle, nil)
```
Comment (andrew-d) [inline ipn/ipnlocal/local.go:3595]: Hm, technically these `defer`s won't fire until after the `for` loop is finished... probably fine in practice, though.

**Review by andrew-d — APPROVED** [review]

File: ipn/ipnlocal/local.go (hunk 3 above)
Comment (bradfitz) [reply ipn/ipnlocal/local.go:3595]: Yeah. I couldn't convince myself it mattered.

**Pull request merged**